<launch>

<!--
  <param name="ticks_meter" value="14865" />
  <param name="encoder_min" value="-2147483648" />
  <param name="encoder_max" value="2147483648" />
-->

  <arg name="model" default="robot1.urdf" />
  <arg name="gui" default="False" />

  <!-- set these parameters on Parameter Server -->
  <param name="robot_description" textfile="$(find rospibot6)/urdf/$(arg model)" />
  <param name="use_gui" value="$(arg gui)"/>

  <rosparam param="base_width">0.182</rosparam>
  <rosparam param="ticks_meter">4897</rosparam>

  <!--<include file="$(find rospibot6)/launch/diff_tf.launch"/>-->
  <node name="diff_tf" pkg="rospibot6" type="diff_tf.py" output="screen">
    <rosparam param="ticks_meter">4897</rosparam>
    <rosparam param="base_width">0.182</rosparam>
    <rosparam param="base_frame_id">base_link</rosparam>
    <rosparam param="odom_frame_id">odom</rosparam>
    <rosparam param="rate">50</rosparam>
  </node>

<!--
  <node name="diff_twist" pkg="rospibot6" type="twist_to_motors.py" output="screen" >
    <rosparam param="base_width">0.182</rosparam>
    <rosparam param="timeout_ticks">2</rosparam>
    <rosparam param="rate">10</rosparam>
  </node>

  <node name="lpid_velocity" pkg="rospibot6" type="pid_velocity.py" output="screen">
    <remap from="wheel" to="lwheel"/>
    <remap from="motor_cmd" to="motor_left"/>
    <remap from="wheel_vtarget" to="lwheel_vtarget"/>
    <remap from="wheel_vel" to="lwheel_vel"/>
    <rosparam param="ticks_meter">4897</rosparam>
    <rosparam param="Kp">30000.0</rosparam>
    <rosparam param="Ki">20000.0</rosparam>
    <rosparam param="Kd">0.0</rosparam>
    <rosparam param="out_min">-32000</rosparam>
    <rosparam param="out_max">32000</rosparam>
    <rosparam param="rate">50</rosparam>
    <rosparam param="timeout_ticks">3</rosparam>
    <rosparam param="vel_threshold">0.001</rosparam>
    <rosparam param="rolling_pts">4</rosparam>
  </node>

  <node name="rpid_velocity" pkg="rospibot6" type="pid_velocity.py" output="screen">
    <remap from="wheel" to="rwheel"/>
    <remap from="motor_cmd" to="motor_right"/>
    <remap from="wheel_vtarget" to="rwheel_vtarget"/>
    <remap from="wheel_vel" to="rwheel_vel"/>
    <rosparam param="ticks_meter">4897</rosparam>
    <rosparam param="Kp">30000.0</rosparam>
    <rosparam param="Ki">20000.0</rosparam>
    <rosparam param="Kd">0.0</rosparam>
    <rosparam param="out_min">-32000</rosparam>
    <rosparam param="out_max">32000</rosparam>
    <rosparam param="rate">50</rosparam>
    <rosparam param="timeout_ticks">3</rosparam>
    <rosparam param="vel_threshold">0.001</rosparam>
    <rosparam param="rolling_pts">4</rosparam>
  </node>
-->
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" />

  <node name="xv11" pkg="xv_11_laser_driver" type="neato_laser_publisher" respawn="true" output="screen">
    <param name="port" value="/dev/ttyAMA0"/>
    <param name="firmware_version" value="2"/>
  </node>

  <node name="gmap" pkg="gmapping" type="slam_gmapping" respawn="true" respawn_delay="30" output="screen">
    <param name="linearUpdate" type="double" value="0.0" />
    <param name="angularUpdate" type="double" value="0.0" />
    <param name="map_update_interval" type="double" value="5.0" />
    <param name="throttle_scans" type="int" value="1" />

    <!-- laser config -->
    <!--
    <param name="maxUrange" type="double" value="10.0" />
    <param name="maxRange" type="double" value="11.0" />
    -->

    <!-- map size & resolution-->
    <param name="xmin" type="double" value="-20.0" />
    <param name="ymin" type="double" value="-20.0" />
    <param name="xmax" type="double" value="20.0" />
    <param name="ymax" type="double" value="20.0" />
    <param name="delta" type="double" value="0.05" />

    <!-- odometry error -->
    <param name="srr" type="double" value="0.1" />
    <param name="srt" type="double" value="0.2" />
    <param name="str" type="double" value="0.1" />
    <param name="stt" type="double" value="0.2" />
  </node>

  <node name="move_base" pkg="move_base" type="move_base" respawn="false" output="screen">
    <rosparam file="$(find rospibot6)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find rospibot6)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find rospibot6)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find rospibot6)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find rospibot6)/param/base_local_planner_params.yaml" command="load" />
    <rosparam file="$(find rospibot6)/param/dwa_local_planner_params.yaml" command="load" />
    <rosparam file="$(find rospibot6)/param/move_base_params.yaml" command="load" />
  </node>

<!--
  <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
    <param name="output_frame" value="odom"/>
    <param name="freq" value="30.0"/>
    <param name="sensor_timeout" value="1.0"/>
    <param name="odom_used" value="true"/>
    <param name="imu_used" value="true"/>
    <param name="vo_used" value="false"/>
    <param name="debug" value="false"/>
    <param name="self_diagnose" value="false"/>
  </node>
-->

  <node name="base_node" pkg="rospibot6" type="base5"  respawn="true" respawn_delay="30" output="screen">
    <!--<rosparam file="$(find mypackage)/config/example.yaml" command="load" />-->
    <remap from="wheel_left" to="lwheel"/>
    <remap from="wheel_right" to="rwheel"/>
    <param name="frequency" type="int" value="100" />
    <param name="pid_enabled" type="bool" value="0" />
    <param name="odom_enabled" type="bool" value="0" />
    <param name="cmdvel_enabled" type="bool" value="1" />
    <param name="cmdvel_k" type="double" value="0.5" />
    <!-- BASE -->
    <param name="wheel_rad" type="double" value="0.0312" />
    <param name="wheel_sep" type="double" value="0.182" />
    <param name="wheel_encoder_pulses" type="double" value="960.0" />
    <!-- PID -->
    <param name="pid_out_trh" type="double" value="0.0" />
    <param name="pid_out_min" type="double" value="0.0" />
    <param name="pid_out_max" type="double" value="64.0" />
    <param name="pid_out_k" type="double" value="500.0" />
    <param name="pid_sample_time" type="double" value="0.050" />
    <param name="pid_setpoint_max" type="double" value="150000.0" />
    <param name="pid_left_kp" type="double" value="0.7" />
    <param name="pid_left_ki" type="double" value="0.01" />
    <param name="pid_left_kd" type="double" value="0.0" />
    <param name="pid_right_kp" type="double" value="0.7" />
    <param name="pid_right_ki" type="double" value="0.01" />
    <param name="pid_right_kd" type="double" value="0.0" />
    <!-- IMU -->
    <param name="frame_id" type="str" value="base_imu" />
    <param name="ax" type="int" value="0" />
    <param name="ay" type="int" value="0" />
    <param name="az" type="int" value="0" />
    <param name="gx" type="int" value="0" />
    <param name="gy" type="int" value="0" />
    <param name="gz" type="int" value="0" />
    <param name="ado" type="bool" value="false" />
    <param name="debug" type="bool" value="false" />
  </node>

</launch>
